Marlin/src/feature/repeat.cpp

```diff
@@ -0,0 +1,69 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#include "../inc/MarlinConfig.h"
+
+#if ENABLED(GCODE_REPEAT_MARKERS)
+
+#include "repeat.h"
+
+#include "../gcode/gcode.h"
+#include "../sd/cardreader.h"
+
+repeat_marker_t Repeat::marker[MAX_REPEAT_NESTING];
+int8_t Repeat::total;
+
+void Repeat::add_marker(const uint32_t sdpos, const uint16_t count) {
+  if (total >= MAX_REPEAT_NESTING)
+    SERIAL_ECHO_MSG("!Too many markers.");
+  else {
+    marker[total].sdpos = sdpos;
+    marker[total].counter = count ?: -1;
```

Comment:
This looks strange. I wouldn't have thought this would even compile. Is this just a ternary operator with no first value? It is hard for me to interpret the intent from this syntax.

Reply:
I see it highlighted multiple lines. This is the one that confuses me:
`    marker[total].counter = count ?: -1;`

Reply:
Set `.counter` to `count` unless it's zero, in which case set it to -1.

Reply:
This tripped me up because it isn't standard C++, it is a GNU extension.
Personally, I would prefer the C++ standard form, but I guess it doesn't matter if we only ever build with GCC.